File: src/main/java/spark/java/TotalSpentByCustomerSorted.java
package spark.java;

import org.apache.log4j.Level;
import org.apache.log4j.Logger;
import org.apache.spark.api.java.*;
import org.apache.spark.api.java.function.*;
import scala.Tuple2;
import java.util.List;

public class TotalSpentByCustomerSorted {

    public static void main(String[] args) {
        Logger.getLogger("org").setLevel(Level.ERROR);
        JavaSparkContext sc = new JavaSparkContext("local[*]", "TotalSpentByCustomerSorted");
        JavaRDD<String> lines = sc.textFile("./data/customer-orders.csv");
        JavaPairRDD<Integer, Float> customerSpent = lines.mapToPair((PairFunction<String, Integer, Float>) line -> {
            String[] fields = line.split(",");
            int customerID = Integer.parseInt(fields[0]);
            float amount = Float.parseFloat(fields[2]);
            return new Tuple2<>(customerID, amount);
        }).reduceByKey(Float::sum);
        JavaPairRDD<Float, Integer> customerSpentSorted = customerSpent.mapToPair(s -> new Tuple2<>(s._2(), s._1())).sortByKey(false, 1);
        List<Tuple2<Float, Integer>> results = customerSpentSorted.collect();
        results.forEach(x -> System.out.println("Customer " + x._2() + ": " + String.format("%.2f", x._1())));
    }

}

File: src/main/scala/spark/scala/TotalSpentByCustomerSorted.scala
package spark.scala

import org.apache.spark._
import org.apache.log4j._

object TotalSpentByCustomerSorted {

  def parseLine(line: String): (Int, Float) = {
    val fields = line.split(",")
    val customerID = fields(0).toInt
    val amount = fields(2).toFloat
    (customerID, amount)
  }

  def main(args: Array[String]): Unit = {
    Logger.getLogger("org").setLevel(Level.ERROR)
    val sc = new SparkContext("local[*]", "TotalSpentByCustomer")
    val lines = sc.textFile("./data/customer-orders.csv")
    val customerSpent = lines.map(parseLine).reduceByKey((v1, v2) => v1 + v2)
    val customerSpentSorted = customerSpent.map(x => (x._2, x._1)).sortByKey(ascending = false, 1)
    val results = customerSpentSorted.collect()
    results.foreach(x => println(f"Customer ${x._2}: ${x._1}%.2f"))
  }

}
